File: app/src/main/java/io/github/gmazzo/android/livewallpaper/weather/engine/particles/ParticleSnow.java
package io.github.gmazzo.android.livewallpaper.weather.engine.particles;

import static javax.microedition.khronos.opengles.GL10.GL_ONE_MINUS_SRC_ALPHA;
import static javax.microedition.khronos.opengles.GL10.GL_SRC_ALPHA;

import javax.microedition.khronos.opengles.GL10;
import javax.microedition.khronos.opengles.GL11;

import io.github.gmazzo.android.livewallpaper.weather.IsolatedRenderer;
import io.github.gmazzo.android.livewallpaper.weather.R;
import io.github.gmazzo.android.livewallpaper.weather.engine.GlobalRand;
import io.github.gmazzo.android.livewallpaper.weather.engine.Vector;
import io.github.gmazzo.android.livewallpaper.weather.engine.scenes.Scene;
import io.github.gmazzo.android.livewallpaper.weather.engine.scenes.SceneBase;
import io.github.gmazzo.android.livewallpaper.weather.engine.scenes.SceneSnow;

public class ParticleSnow extends ParticleSystem {
    public ParticleSnow() {
        this.spawnRate = 0.25f;
        this.spawnRateVariance = 0.05f;
        this.startEngineColor.set(1.0f, 1.0f, 1.0f, 3.0f);
        this.destEngineColor.set(1.0f, 1.0f, 1.0f, 0.0f);
        this.spawnRangeX = 20.0f;
    }

    @Override
    public void render(GL11 gl, Vector systemOrigin) {
        if (model == null) {
            model = Scene.sModels.get(R.raw.flakes);
            texture = Scene.sTextures.get(R.raw.p_snow1);
        }
        super.render(gl, systemOrigin);
    }

    public void particleSetup(Particle particle) {
        super.particleSetup(particle);
        float bias = ((IsolatedRenderer.homeOffsetPercentage * 2.0f) - 1.0f) * 4.0f;
        particle.lifetime = 4.5f;
        particle.startScale.set(GlobalRand.floatRange(0.15f, 0.3f));
        particle.destScale.set(GlobalRand.floatRange(0.15f, 0.3f));
        float randX1 = (GlobalRand.floatRange(-6.0f, 6.0f) * SceneSnow.pref_snowNoise) + bias;
        float randX2 = (GlobalRand.floatRange(-8.0f, 8.0f) * SceneSnow.pref_snowNoise) + bias;
        float randY1 = GlobalRand.floatRange(-2.0f, 2.0f);
        float randY2 = GlobalRand.floatRange(-2.0f, 2.0f);
        float randZ = GlobalRand.floatRange(-(3.0f + (SceneSnow.pref_snowGravity * 1.5f)), -3.0f);
        particle.startVelocity.set(randX1, randY1, randZ);
        particle.destVelocity.set(randX2, randY2, randZ);
    }

    public void renderEnd(GL10 gl) {
    }

    public void renderStart(GL10 gl) {
        gl.glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
    }

    public void update(float timeDelta) {
        super.update(timeDelta);
        this.startEngineColor.set(SceneBase.todEngineColorFinal.getR(), SceneBase.todEngineColorFinal.getG(), SceneBase.todEngineColorFinal.getB(), 3.0f);
        this.destEngineColor.set(SceneBase.todEngineColorFinal.getR(), SceneBase.todEngineColorFinal.getG(), SceneBase.todEngineColorFinal.getB(), 0.0f);
    }
}

File: app/src/main/kotlin/io/github/gmazzo/android/livewallpaper/weather/engine/particles/ParticleSnow.kt
package io.github.gmazzo.android.livewallpaper.weather.engine.particles

import io.github.gmazzo.android.livewallpaper.weather.R
import io.github.gmazzo.android.livewallpaper.weather.WeatherViewRenderer.Companion.homeOffsetPercentage
import io.github.gmazzo.android.livewallpaper.weather.engine.models.Models
import io.github.gmazzo.android.livewallpaper.weather.engine.nextFloat
import io.github.gmazzo.android.livewallpaper.weather.engine.scenes.Scene
import io.github.gmazzo.android.livewallpaper.weather.engine.scenes.SceneSnow
import io.github.gmazzo.android.livewallpaper.weather.engine.textures.Textures
import javax.inject.Inject
import javax.microedition.khronos.opengles.GL10
import kotlin.random.Random

class ParticleSnow @Inject constructor(
    models: Models,
    textures: Textures,
) : ParticleSystem(models[R.raw.flakes], textures[R.raw.p_snow1]) {

    init {
        this.spawnRate = 0.25f
        this.spawnRateVariance = 0.05f
        startEngineColor.set(1.0f, 1.0f, 1.0f, 3.0f)
        destEngineColor.set(1.0f, 1.0f, 1.0f, 0.0f)
        this.spawnRangeX = 20.0f
    }

    override fun particleSetup(particle: Particle?) {
        super.particleSetup(particle)
        val bias: Float = ((homeOffsetPercentage * 2.0f) - 1.0f) * 4.0f
        particle!!.lifetime = 4.5f
        particle.startScale.set(Random.nextFloat(0.15f, 0.3f))
        particle.destScale.set(Random.nextFloat(0.15f, 0.3f))
        val randX1: Float =
            (Random.nextFloat(-6.0f, 6.0f) * SceneSnow.Companion.pref_snowNoise) + bias
        val randX2: Float =
            (Random.nextFloat(-8.0f, 8.0f) * SceneSnow.Companion.pref_snowNoise) + bias
        val randY1 = Random.nextFloat(-2.0f, 2.0f)
        val randY2 = Random.nextFloat(-2.0f, 2.0f)
        val randZ =
            Random.nextFloat(-(3.0f + (SceneSnow.Companion.pref_snowGravity * 1.5f)), -3.0f)
        particle.startVelocity.set(randX1, randY1, randZ)
        particle.destVelocity.set(randX2, randY2, randZ)
    }

    public override fun renderEnd(gl: GL10?) {
    }

    public override fun renderStart(gl: GL10) {
        gl.glBlendFunc(GL10.GL_SRC_ALPHA, GL10.GL_ONE_MINUS_SRC_ALPHA)
    }

    override fun update(timeDelta: Float) {
        super.update(timeDelta)
        startEngineColor.set(
            Scene.todEngineColorFinal!!.r,
            Scene.todEngineColorFinal!!.g,
            Scene.todEngineColorFinal!!.b,
            3.0f
        )
        destEngineColor.set(
            Scene.todEngineColorFinal!!.r,
            Scene.todEngineColorFinal!!.g,
            Scene.todEngineColorFinal!!.b,
            0.0f
        )
    }
}
